src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler.cs

```diff
@@ -0,0 +1,150 @@
+
+        public override async Task Write(T value, NpgsqlWriteBuffer buffer, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
+        {
+            if (buffer.WriteSpaceLeft < sizeof(int))
+                await buffer.Flush(async);
+
+            buffer.WriteInt32(_members.Length);
+
+            foreach (var member in _members)
+                await member.Write(value, buffer, lengthCache, async);
+        }
+
+        public override int ValidateAndGetLength(T value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
+        {
+            if (lengthCache == null)
+                lengthCache = new NpgsqlLengthCache(1);
+
+            if (lengthCache.IsPopulated)
+                return lengthCache.Get();
+
+            // Leave empty slot for the entire composite type, and go ahead an populate the element slots
+            var position = lengthCache.Position;
+            lengthCache.Set(0);
+
+            // number of fields + (type oid + field length) * member count
+            var length = sizeof(int) + sizeof(int) * 2 * _members.Length;
+            foreach (var member in _members)
+                length += member.ValidateAndGetLength(value, ref lengthCache);
+
+            return lengthCache.Lengths[position] = length;
+        }
+
+        public static MappedCompositeHandler<T> Create(PostgresCompositeType pgType, ConnectorTypeMapper typeMapper, INpgsqlNameTranslator nameTranslator)
+        {
+            var clrType = typeof(T);
+            var pgFields = pgType.Fields;
+            var clrHandlers = new CompositeMemberHandler<T>[pgFields.Count];
+            var clrHandlerCount = 0;
+
+            var clrHandlerType = IsValueType<T>.Value
+                ? typeof(CompositeMemberHandlerOfStruct<,>)
+                : typeof(CompositeMemberHandlerOfClass<,>);
+
+            foreach (var clrMember in clrType.GetMembers(BindingFlags.Instance | BindingFlags.Public))
+            {
+                Type clrMemberType;
+                switch (clrMember)
+                {
+                    case FieldInfo clrField:
+                        clrMemberType = clrField.FieldType;
+                        break;
+                    case PropertyInfo clrProperty:
+                        clrMemberType = clrProperty.PropertyType;
+                        break;
+                    default:
+                        continue;
+                }
+                
+                var attr = clrMember.GetCustomAttribute<PgNameAttribute>();
+                var name = attr != null
+                    ? attr.PgName
+                    : nameTranslator.TranslateMemberName(clrMember.Name);
+
+                int pgFieldIndex;
+                PostgresCompositeType.Field pgField = null;
+
+                for (pgFieldIndex = pgFields.Count - 1; pgFieldIndex >= 0; --pgFieldIndex)
+                {
+                    pgField = pgFields[pgFieldIndex];
+                    if (pgField.Name == name) break;
```

Comment:
_This isn't my area of expertise, so please forgive what may actually be a silly question:_

Is there a perf reason to avoid checking `pgFields[pgFieldIndex].Name == name` directly? Is repeatedly reassigning the local cheaper than the one additional index access?

For example, assuming that we're dealing with a composite of 4 fields, `(a, b, c, d)` and we want to get field `a`:

The current version (`pgField.Name == name`):
- indexes `pgFields` 4 times
- accesses `.Name` 4 times
- writes `pgField` 4 times

But if it checked `pgFields[pgFieldIndex].Name == name` directly, it would need to:
- indexes `pgFields` 5 times
- accesses `.Name` 4 times 
- writes `pgField` 1 time.

Reply:
Don't know and will check.

Reply:
Checked and verified that a local variable could improve perf when an iteration goes over an array. Probably it's the right behavior because there could be side effects and the JIT takes that into account. The last two tests (without direct return) demonstrate the current scenario when the result is used after a cycle.

One small note before you will see the test results. This isn't applicable to the commented code since `pgFields` isn't an array but an `IReadOnlyList<T>`. Therefore, the JIT must call the indexer each time.

```
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.16299.1029 (1709/FallCreatorsUpdate/Redstone3)
Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
Frequency=2062499 Hz, Resolution=484.8487 ns, Timer=TSC
.NET Core SDK=2.2.104
  [Host]     : .NET Core 2.2.2 (CoreCLR 4.6.27317.07, CoreFX 4.6.27318.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.2 (CoreCLR 4.6.27317.07, CoreFX 4.6.27318.02), 64bit RyuJIT
```

|                                Method |     Mean |     Error |    StdDev |
|-------------------------------------- |---------:|----------:|----------:|
|      Iteration_WithLocal_DirectReturn | 1.972 ns | 0.1022 ns | 0.1177 ns |
|   Iteration_WithoutLocal_DirectReturn | 2.138 ns | 0.1012 ns | 0.1039 ns |
|    Iteration_WithLocal_NoDirectReturn | 1.713 ns | 0.0952 ns | 0.0978 ns |
| Iteration_WithoutLocal_NoDirectReturn | 1.956 ns | 0.1051 ns | 0.1438 ns |

```csharp
private readonly Field[] pgFields = new[]
{
    new Field("one"),
    new Field("two"),
    new Field("three"),
};

[Benchmark]
public Field Iteration_WithLocal_DirectReturn()
{
    for (var pgFieldIndex = pgFields.Length - 1; pgFieldIndex >= 0; --pgFieldIndex)
    {
        var pgField = pgFields[pgFieldIndex];
        if (pgField.Name == "three")
        {
            return pgField;
        }
    }

    return null;
}

[Benchmark]
public Field Iteration_WithoutLocal_DirectReturn()
{
    for (var pgFieldIndex = pgFields.Length - 1; pgFieldIndex >= 0; --pgFieldIndex)
    {
        if (pgFields[pgFieldIndex].Name == "three")
        {
            return pgFields[pgFieldIndex];
        }
    }

    return null;
}

[Benchmark]
public Field Iteration_WithLocal_NoDirectReturn()
{
    Field pgField = null;

    for (var pgFieldIndex = pgFields.Length - 1; pgFieldIndex >= 0; --pgFieldIndex)
    {
        pgField = pgFields[pgFieldIndex];

        if (pgField.Name == "three")
        {
            break;
        }
    }

    return pgField;
}

[Benchmark]
public Field Iteration_WithoutLocal_NoDirectReturn()
{
    Field pgField = null;

    for (var pgFieldIndex = pgFields.Length - 1; pgFieldIndex >= 0; --pgFieldIndex)
    {
        if (pgFields[pgFieldIndex].Name == "three")
        {
            pgField = pgFields[pgFieldIndex];
            break;
        }
    }

    return pgField;
}
```

Reply:
Very interesting, thanks for taking a look at it!